File: backend/app/routers/checkins.py
from datetime import datetime
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, delete
from app.database import get_db
from app.models.orm import CheckInToken, CheckInResponse, Farm
from app.schemas.schemas import TokenValidation, QuestionOut, SubmitIn
from app.services.magic_link import validate_token
from app.services.questions import get_questions_for_farm

router = APIRouter()


async def _get_token_or_404(token_str: str, db: AsyncSession) -> CheckInToken:
    token = await validate_token(db, token_str)
    if token is None:
        raise HTTPException(status_code=404, detail="Link not found or expired")
    return token


@router.get("/{token}", response_model=TokenValidation)
async def get_checkin(token: str, db: AsyncSession = Depends(get_db)):
    tok = await _get_token_or_404(token, db)

    farm_result = await db.execute(select(Farm).where(Farm.id == tok.farm_id))
    farm = farm_result.scalar_one()

    month = tok.period_start.month
    questions = get_questions_for_farm(farm.enterprise_types, month)

    # Load existing responses for this token
    resp_result = await db.execute(
        select(CheckInResponse).where(CheckInResponse.token_id == tok.id)
    )
    existing = resp_result.scalars().all()
    existing_map: dict[str, str | float | None] = {}
    for r in existing:
        val = r.value_num if r.value_num is not None else (r.value_option or r.value_text)
        existing_map[r.question_key] = val

    return TokenValidation(
        farm=farm,
        period_start=tok.period_start,
        period_end=tok.period_end,
        questions=[
            QuestionOut(
                key=q.key,
                section=q.section,
                label=q.label,
                type=q.type,
                unit=q.unit,
                options=q.options or None,
                hint=q.hint,
            )
            for q in questions
        ],
        existing_responses=existing_map,
        completed=tok.completed_at is not None,
    )


@router.post("/{token}/save")
async def save_response(token: str, body: SubmitIn, db: AsyncSession = Depends(get_db)):
    """Upsert individual responses (partial save — farmer can return later)."""
    tok = await _get_token_or_404(token, db)

    for resp in body.responses:
        # Delete existing response for this question in this token period
        await db.execute(
            delete(CheckInResponse).where(
                CheckInResponse.token_id == tok.id,
                CheckInResponse.question_key == resp.question_key,
            )
        )
        db.add(CheckInResponse(
            farm_id=tok.farm_id,
            token_id=tok.id,
            period_start=tok.period_start,
            section=resp.section,
            question_key=resp.question_key,
            value_num=resp.value_num,
            value_text=resp.value_text,
            value_option=resp.value_option,
            submitted_at=datetime.utcnow(),
        ))

    await db.commit()
    return {"saved": len(body.responses)}


@router.post("/{token}/complete")
async def complete_checkin(token: str, body: SubmitIn, db: AsyncSession = Depends(get_db)):
    """Save final responses and mark token as completed."""
    tok = await _get_token_or_404(token, db)

    for resp in body.responses:
        await db.execute(
            delete(CheckInResponse).where(
                CheckInResponse.token_id == tok.id,
                CheckInResponse.question_key == resp.question_key,
            )
        )
        db.add(CheckInResponse(
            farm_id=tok.farm_id,
            token_id=tok.id,
            period_start=tok.period_start,
            section=resp.section,
            question_key=resp.question_key,
            value_num=resp.value_num,
            value_text=resp.value_text,
            value_option=resp.value_option,
            submitted_at=datetime.utcnow(),
        ))

    tok.completed_at = datetime.utcnow()
    await db.commit()
    return {"completed": True}

File: worker/src/routes/checkins.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";
import { getDb } from "../db/supabase";
import { getQuestionsForFarm } from "../services/questions";
import { isExpired } from "../services/magic_link";

type Bindings = {
  SUPABASE_URL: string;
  SUPABASE_SERVICE_ROLE_KEY: string;
  APP_BASE_URL: string;
};

const router = new Hono<{ Bindings: Bindings }>();

const ResponseSchema = z.object({
  question_key: z.string(),
  section: z.string(),
  value_num: z.number().nullable().optional(),
  value_text: z.string().nullable().optional(),
  value_option: z.string().nullable().optional(),
});

const SubmitSchema = z.object({ responses: z.array(ResponseSchema) });

// GET /api/checkins/:token
router.get("/:token", async (c) => {
  const db = getDb(c.env);

  const { data: tok } = await db
    .from("check_in_tokens")
    .select("*")
    .eq("token", c.req.param("token"))
    .single();

  if (!tok) return c.json({ error: "Link not found or expired" }, 404);
  if (isExpired(tok.expires_at)) return c.json({ error: "Link not found or expired" }, 404);

  // Record first access
  if (!tok.first_accessed_at) {
    await db.from("check_in_tokens")
      .update({ first_accessed_at: new Date().toISOString() })
      .eq("id", tok.id);
  }

  const { data: farm } = await db.from("farms").select("*").eq("id", tok.farm_id).single();
  if (!farm) return c.json({ error: "Farm not found" }, 404);

  const month = new Date(tok.period_start).getMonth() + 1;
  const questions = getQuestionsForFarm(farm.enterprise_types, month);

  const { data: existing } = await db
    .from("check_in_responses")
    .select("question_key, value_num, value_text, value_option")
    .eq("token_id", tok.id);

  const existingMap: Record<string, string | number | null> = {};
  for (const r of existing ?? []) {
    existingMap[r.question_key] = r.value_num ?? r.value_option ?? r.value_text;
  }

  return c.json({
    farm,
    period_start: tok.period_start,
    period_end: tok.period_end,
    questions: questions.map((q) => ({
      key: q.key, section: q.section, label: q.label,
      type: q.type, unit: q.unit ?? null, options: q.options ?? null, hint: q.hint ?? null,
    })),
    existing_responses: existingMap,
    completed: tok.completed_at !== null,
  });
});

// POST /api/checkins/:token/save
router.post("/:token/save", zValidator("json", SubmitSchema), async (c) => {
  const db = getDb(c.env);

  const { data: tok } = await db
    .from("check_in_tokens").select("*").eq("token", c.req.param("token")).single();
  if (!tok || isExpired(tok.expires_at)) return c.json({ error: "Link not found or expired" }, 404);

  const { responses } = c.req.valid("json");
  const now = new Date().toISOString();

  for (const r of responses) {
    // Delete then re-insert to achieve upsert on (token_id, question_key)
    await db.from("check_in_responses")
      .delete()
      .eq("token_id", tok.id)
      .eq("question_key", r.question_key);

    await db.from("check_in_responses").insert({
      farm_id: tok.farm_id,
      token_id: tok.id,
      period_start: tok.period_start,
      section: r.section,
      question_key: r.question_key,
      value_num: r.value_num ?? null,
      value_text: r.value_text ?? null,
      value_option: r.value_option ?? null,
      submitted_at: now,
    });
  }

  return c.json({ saved: responses.length });
});

// POST /api/checkins/:token/complete
router.post("/:token/complete", zValidator("json", SubmitSchema), async (c) => {
  const db = getDb(c.env);

  const { data: tok } = await db
    .from("check_in_tokens").select("*").eq("token", c.req.param("token")).single();
  if (!tok || isExpired(tok.expires_at)) return c.json({ error: "Link not found or expired" }, 404);

  const { responses } = c.req.valid("json");
  const now = new Date().toISOString();

  for (const r of responses) {
    await db.from("check_in_responses")
      .delete().eq("token_id", tok.id).eq("question_key", r.question_key);

    await db.from("check_in_responses").insert({
      farm_id: tok.farm_id,
      token_id: tok.id,
      period_start: tok.period_start,
      section: r.section,
      question_key: r.question_key,
      value_num: r.value_num ?? null,
      value_text: r.value_text ?? null,
      value_option: r.value_option ?? null,
      submitted_at: now,
    });
  }

  await db.from("check_in_tokens")
    .update({ completed_at: now }).eq("id", tok.id);

  return c.json({ completed: true });
});

export default router;
